pkg/reconciler/route/resources/virtual_service.go

```diff
@@ -34,13 +37,40 @@ const (
 )
 
 // VirtualServiceName gets the name of a VirtualService given the route.
-func VirtualServiceName(hostname, domain, urlPath string) string {
-	return strconv.FormatUint(
+func VirtualServiceName(parts ...string) string {
+	prefix := strings.Join(parts, "-")
+
+	// Base 36 uses the characters 0-9a-z. The maximum number of chars it
+	// requires is 13 for a Uin64.
+	checksum := strconv.FormatUint(
 		crc64.Checksum(
-			[]byte(hostname+domain+path.Join("/", urlPath)),
+			[]byte(strings.Join(parts, "")),
 			crc64.MakeTable(crc64.ECMA),
 		),
-		10)
+		36)
+
+	prefix = strings.ToLower(prefix)
+
+	// Remove all non-alphanumeric characters.
+	prefix = regexp.MustCompile(`[^a-z0-9-_]`).ReplaceAllString(prefix, "-")
+
+	// First char must be alphanumeric.
+	for len(prefix) > 0 && regexp.MustCompile(`[^a-z0-9]`).Match([]byte{prefix[0]}) {
```

Comment:
Same comment as above

Reply:
The prefix may still have non-alphanumeric characters whereas the hash will only have 0-9a-z. We want to prepend the hostname and domain (and path in the case of a route). These things can contain illegal characters like a '/'.